src/index.js

```diff
@@ -95,48 +111,195 @@ class Masq {
+   */
+  exchangeDataHyperdbKeys () {
+    // Subscribe to channel for a limited time to sync with masq
+    if (!this.profile) throw (new Error('No profile selected'))
+    const hub = signalhub(this.channel, [HUB_URL])
+    let sw = null
 
     if (swarm.WEBRTC_SUPPORT) {
-      this.sw = swarm(this.hub)
+      sw = swarm(hub)
     } else {
-      this.sw = swarm(this.hub, { wrtc: require('wrtc') })
+      sw = swarm(hub, { wrtc: require('wrtc') })
     }
 
-    this.sw.on('peer', (peer, id) => {
+    sw.on('peer', (peer, id) => {
       // check challenges
-      peer.on('data', data => this._handleData(data))
+      peer.on('data', data => _handleData(data, peer))
+    })
+
+    sw.on('close', () => {
+      hub.close()
+    })
+
+    sw.on('disconnect', (peer, id) => {
+      sw.close()
+      hub.close()
+    })
+
+    const _handleData = async (data, peer) => {
+      const json = JSON.parse(data)
+
+      switch (json.msg) {
+        case 'sendDataKey':
+          if (json.challenge !== this.challenge) {
+            // This peer may be malicious, close the connection
+            sw.close()
+            hub.close()
+          } else {
+            // db creation and replication
+            const db = hyperdb(rai(this.profile), Buffer.from(json.key, 'hex'), { valueEncoding: 'json' })
+            await dbReady(db)
+            // Store
+            this.dbs[this.profile] = db
+
+            peer.send(JSON.stringify({
+              msg: 'requestWriteAccess',
+              key: db.local.key.toString('hex')
+            }))
+          }
+          break
+        case 'ready':
+          // Masq must send ready after the authorization
+          this._startDataReplication()
+          sw.close()
+          break
+        default:
+          break
+      }
+    }
+  }
+
+  async _startReplication (db) {
+    const discoveryKey = db.discoveryKey.toString('hex')
+    this.hubs['syncProfiles'] = signalhub(discoveryKey, [HUB_URL])
+    const hub = this.hubs['syncProfiles']
+
+    if (swarm.WEBRTC_SUPPORT) {
+      this.sws['syncProfiles'] = swarm(hub)
+    } else {
+      this.sws['syncProfiles'] = swarm(hub, { wrtc: require('wrtc') })
+    }
+    const sw = this.sws['syncProfiles']
+
+    sw.on('peer', async (peer, id) => {
+      const stream = db.replicate({ live: true })
+      pump(peer, stream, peer)
     })
 
-    this.sw.on('close', () => {
-      this.hub.close()
+    sw.on('close', () => {
+      hub.close()
     })
 
-    this.sw.on('disconnect', (peer, id) => {
-      this.sw.close()
-      this.hub.close()
+    sw.on('disconnect', (peer, id) => {
+      sw.close()
+      hub.close()
     })
   }
 
-  _handleData (data) {
-    const json = JSON.parse(data)
-    switch (json.msg) {
-      case 'challenge':
-        if (json.challenge !== this.challenge) {
-          // This peer may be malicious, close the connection
-          this.sw.close()
-          this.hub.close()
-        }
-        break
+  async _startDataReplication () {
```

Comment:
_startReplication and _startDataReplication seems to share 99% of code, can we refactor to only one function ?

Reply:
yes, done